graphql-protocol/src/main/java/com/linecorp/armeria/internal/server/graphql/protocol/GraphqlUtil.java

```diff
@@ -43,33 +42,31 @@ public final class GraphqlUtil {
      * specification</a>.
      */
     @Nullable
-    public static MediaType produceType(RequestHeaders headers) {
-        final MediaType contentType = headers.contentType();
-        if (HttpMethod.POST == headers.method() && contentType != null) {
-            if (contentType.is(MediaType.GRAPHQL) ||
-                contentType.is(MediaType.MULTIPART_FORM_DATA)) {
-                return MediaType.GRAPHQL_JSON;
-            }
-        }
-
+    public static MediaType produceType(RequestHeaders headers, MediaType defaultProduceType) {
```

Comment:
Should we take a function to select a response type dynamically which would be useful to handle both legacy and new clients in a server?
```java
public static MediaType produceType(ServiceRequestContext ctx, Function<ServiceRequestContext, MediaType> defaultProduceTypeFunction) {
     ...
}

...
GrpcService.builder()
           // statically set 
           .defaultProduceType(MediaType.GRAPHQL_JSON)
           // or dynamically choose
           .defaultProduceType(ctx -> {
               if (legacyclient) {
                   return MediaType.GRAPHQL_JSON;
               } else {
                   return MediaType.GRAPHQL_RESPONSE_JSON;
               }
           });

```

Reply:
> Should we take a function to select a response type dynamically 

Is there any usage to select the type dynamically except using accept header which we are already handling?

Reply:
`defaultProduceType` is used when `Accept` header is empty. 
As per [`Legacy watershed`](https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#legacy-watershed-1), the response content-type could be either `application/json` or `application/graphql-response+json`.

> Before 1st January 2025 (2025-01-01T00:00:00Z), if the client does not supply an Accept header, the server SHOULD treat the request as if it had Accept: application/json. From 1st January 2025 (2025-01-01T00:00:00Z), if the client does not supply an Accept header, the server SHOULD treat the request as if it had Accept: application/graphql-response+json.

If an `Accept` header is absent, I imagine that a server could choose `application/json` for a legacy client and ` application/graphql-response+json` for a new client in corner cases. 
But I'm not sure there is such a desire for that.

Reply:
> I imagine that a server could choose application/json for a legacy client and application/graphql-response+json for a new client in corner cases.

If it's distinguishable, I think we can add the logic to our code. However, I thought it was hard to determine if the client is the legacy one or not.

Reply:
`User-Agent` of clients may be used. In addition, GraphQL servers could switch to the response type after the deadline.  
If you think it looks overkill, we add it later without breaking changes when there is a demand for it.

Reply:
Had a chat with the team we found out that the current implementation does not respect spec.
If accept is empty, we should return `application/json` before 2025 year according to the spec.
So I will fix the current logic and remove the setter at the moment.